<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Mpd : Multiplex PCR Design">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Mpd</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/wingolab-org/mpd-c">View on GitHub</a>

          <h1 id="project_title">Mpd</h1>
          <h2 id="project_tagline">Multiplex PCR Design</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/wingolab-org/mpd-c/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/wingolab-org/mpd-c/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a id="mpd---multiplex-pcr-design" class="anchor" href="#mpd---multiplex-pcr-design" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>MPD - Multiplex PCR Design</h1>

<p>by Thomas Wingo and David Cutler</p>

<h2>
<a id="description" class="anchor" href="#description" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Description</h2>

<p>MPD is a program designed to automate creation of multiplex primer design written in C. The <code>mpd_lessGreedy</code> and <code>mpd_moreGreedy</code> binaries differ in which primer pool they choose to start with for pool creation. Either binary can be used as stand-alone or in conjunction with the <a href="http://github.com/wingolab-org/mpd-perl">MPD perl package</a>.</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h2>

<ul>
<li>Clone the repository</li>
<li>Make with <code>make all</code>
</li>
<li>Binaries will be compiled and saved to the <code>build</code> directory</li>
</ul>

<h2>
<a id="required-files" class="anchor" href="#required-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Required files</h2>

<ul>
<li>You will need a hashed copy of the genome to run the primer software.</li>
<li>Instructions below show how you can create one yourself. A prebuild hg38 genome with flat dbSnp files is available from <a href="https://bitbucket.org/wingolab/mpd-dat/">this repository</a>. It may be cloned like so, <code>git clone https://bitbucket.org/wingolab/mpd-dat.git</code>.</li>
</ul>

<h3>
<a id="build-hashed-genome" class="anchor" href="#build-hashed-genome" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Build Hashed Genome</h3>

<ul>
<li>Download the genome of interest as a fasta file</li>
<li>Use <code>bin/run_index.pl</code>, which creates a sh script to run <code>index_genome</code>
</li>
</ul>

<h3>
<a id="flat-dbsnp" class="anchor" href="#flat-dbsnp" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Flat dbSnp</h3>

<ul>
<li>These can be obtained from this <a href="https://bitbucket.org/wingolab/mpd-dat/">this repository</a>, which were prepared from dbSNP version 140.</li>
<li>To create your own flat snp file set based on criteria of your own devising, each line should contain tab-delimited fields of the following:</li>
</ul>

<pre><code>name numberOfReporters chrom position MinorAlleleFrequency allele1/allele2
</code></pre>

<ul>
<li>The <code>numberOfReporters</code> field is no longer used but retained for backwards compatibility. </li>
<li>Prepare a <code>sdx</code> file that contains the number of chromsome files to include as the 1st line and then a list of the names of all chromosome files. On the command line you might try: <code>ls -1 *.line | wc -l &gt; db_flat.sdx; ls -1 &gt;&gt; db_flat.sdx</code>. Note that the sdx should be in the same order that the chromsomes are in for the indexed genome. See the genome's sdx file (e.g., <code>cat hg38.d14.sdx</code>) to see the order.</li>
</ul>

<h2>
<a id="run-mpd" class="anchor" href="#run-mpd" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Run mpd</h2>

<ul>
<li>The easiest way of using MPD is to use the <a href="http://github.com/wingolab-org/mpd-perl">Perl pacakge MPD</a>, but either <code>mpd_lessGreedy</code> and <code>mpd_moreGreedy</code> binaries may be executed from the command line interactively.</li>
</ul>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Mpd maintained by <a href="https://github.com/wingolab-org">wingolab-org</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
